README.md
# benchmark fs.appendFile and fs.writeFile with option mkdirRecursive

#### run windows benchmark
1. install git-for-windows

2. open git-bash shell

3. to benchmark patched nodejs, run
    `./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt`

3. to benchmark vanilla nodejs, run
    `./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt`


#### windows benchmark result
```
async-operation                                         master-branch           pr-branch       performance-impact
30 * 1000 * (fs.writeFile      w/ no-mkdir      )      934 +/-  50 ms      929 +/-  82 ms    no performance-impact
30 * 1000 * (fs.writeFile      w/ fs.mkdir      )      722 +/- 163 ms      722 +/- 159 ms    no performance-impact
30 * 1000 * (fs.writeFile      w/ mkdirRecursive)             ---- ms     1244 +/- 135 ms    72% slower (lazy-mkdirRecursive vs eager-fs.mkdir)

async-operation                                         master-branch           pr-branch       performance-impact
30 * 1000 * (fs.appendFile     w/ no-mkdir      )      983 +/-  62 ms      966 +/-  92 ms    no performance-impact
30 * 1000 * (fs.appendFile     w/ fs.mkdir      )      739 +/- 175 ms      730 +/- 184 ms    no performance-impact
30 * 1000 * (fs.appendFile     w/ mkdirRecursive)             ---- ms     1252 +/- 126 ms    69% slower (lazy-mkdirRecursive vs eager-fs.mkdir)


 sync-operation                                         master-branch           pr-branch       performance-impact
30 * 1000 * (fs.writeFileSync  w/ no-mkdir      )     1008 +/-  42 ms     1007 +/-  45 ms    no performance-impact
30 * 1000 * (fs.writeFileSync  w/ fs.mkdirSync  )     1643 +/- 121 ms     1611 +/- 117 ms    no performance-impact
30 * 1000 * (fs.writeFileSync  w/ mkdirRecursive)             ---- ms     1791 +/-  87 ms     9% slower (lazy-mkdirRecursive vs eager-fs.mkdir)

 sync-operation                                         master-branch           pr-branch       performance-impact
30 * 1000 * (fs.appendFileSync w/ no-mkdir      )      999 +/-  47 ms     1019 +/-  53 ms    no performance-impact
30 * 1000 * (fs.appendFileSync w/ fs.mkdirSync  )     1617 +/- 103 ms     1619 +/- 109 ms    no performance-impact
30 * 1000 * (fs.appendFileSync w/ mkdirRecursive)             ---- ms     1789 +/- 104 ms    11% slower (lazy-mkdirRecursive vs eager-fs.mkdir)
```


CHANGELOG.md
# Node.js Changelog

Select a Node.js version below to view the changelog history:

* [Node.js 14](doc/changelogs/CHANGELOG_V14.md) **Long Term Support**
* [Node.js 13](doc/changelogs/CHANGELOG_V13.md) **Current**
* [Node.js 12](doc/changelogs/CHANGELOG_V12.md) **Long Term Support**
* [Node.js 11](doc/changelogs/CHANGELOG_V11.md) End-of-Life
* [Node.js 10](doc/changelogs/CHANGELOG_V10.md) Long Term Support
* [Node.js 9](doc/changelogs/CHANGELOG_V9.md) End-of-Life
* [Node.js 8](doc/changelogs/CHANGELOG_V8.md) End-of-Life
* [Node.js 7](doc/changelogs/CHANGELOG_V7.md) End-of-Life
* [Node.js 6](doc/changelogs/CHANGELOG_V6.md) End-of-Life
* [Node.js 5](doc/changelogs/CHANGELOG_V5.md) End-of-Life
* [Node.js 4](doc/changelogs/CHANGELOG_V4.md) End-of-Life
* [io.js](doc/changelogs/CHANGELOG_IOJS.md) End-of-Life
* [Node.js 0.12](doc/changelogs/CHANGELOG_V012.md) End-of-Life
* [Node.js 0.10](doc/changelogs/CHANGELOG_V010.md) End-of-Life
* [Archive](doc/changelogs/CHANGELOG_ARCHIVE.md)

Please use the following table to find the changelog for a specific Node.js
release.

<!--lint disable maximum-line-length-->
<table>
<tr>
  <th title="LTS Until 2023-04"><a href="doc/changelogs/CHANGELOG_V14.md">14</a><sup>LTS</sup></th>
  <th title="Current"><a href="doc/changelogs/CHANGELOG_V13.md">13</a><sup>Current</sup></th>
  <th title="LTS Until 2022-04"><a href="doc/changelogs/CHANGELOG_V12.md">12</a><sup>LTS</sup></th>
  <th title="LTS Until 2021-04"><a href="doc/changelogs/CHANGELOG_V10.md">10</a><sup>LTS</sup></th>
</tr>
<tr>
    <td valign="top">
<b><a href="doc/changelogs/CHANGELOG_V14.md#14.15.0">14.15.0</a></b><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.14.0">14.14.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.13.1">14.13.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.13.0">14.13.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.12.0">14.12.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.11.0">14.11.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.10.1">14.10.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.10.0">14.10.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.9.0">14.9.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.8.0">14.8.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.7.0">14.7.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.6.0">14.6.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.5.0">14.5.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.4.0">14.4.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.3.0">14.3.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.2.0">14.2.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.1.0">14.1.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V14.md#14.0.0">14.0.0</a><br/>
    </td>
    <td valign="top">
<b><a href="doc/changelogs/CHANGELOG_V12.md#12.16.2">12.16.2</a></b><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.16.1">12.16.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.16.0">12.16.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.15.0">12.15.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.14.1">12.14.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.14.0">12.14.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.13.1">12.13.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.13.0">12.13.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.12.0">12.12.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.11.1">12.11.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.11.0">12.11.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.10.0">12.10.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.9.1">12.9.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.9.0">12.9.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.8.1">12.8.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.8.0">12.8.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.7.0">12.7.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.6.0">12.6.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.5.0">12.5.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.4.0">12.4.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.3.1">12.3.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.3.0">12.3.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.2.0">12.2.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.1.0">12.1.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.0.0">12.0.0</a><br/>
    </td>
    <td valign="top">
<b><a href="doc/changelogs/CHANGELOG_V10.md#10.20.1">10.20.1</a></b><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.20.0">10.20.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.19.0">10.19.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.18.1">10.18.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.18.0">10.18.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.17.0">10.17.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.16.3">10.16.3</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.16.2">10.16.2</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.16.1">10.16.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.16.0">10.16.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.15.3">10.15.3</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.15.2">10.15.2</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.15.1">10.15.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.15.0">10.15.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.14.2">10.14.2</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.14.1">10.14.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.14.0">10.14.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.13.0">10.13.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.12.0">10.12.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.11.0">10.11.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.10.0">10.10.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.9.0">10.9.0</a><br/>

benchmark.sh
#!/bin/sh
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt
./node.mkdirRecursive.exe benchmark.js modeMkdirRecursive | tee -a .benchmark.mkdirRecursive.txt
./node.commit-3f4ec9c3.exe benchmark.js | tee -a .benchmark.vanilla.txt


benchmark.parse.js
/* jslint utility2:true */
(function () {
    "use strict";
    function parse(filename) {
        let dict;
        dict = {};
        require("fs").readFileSync(filename, "utf8").replace((
            /^(.*?)\u0020-\u0020time\u0020elapsed\u0020-\u0020\u0020*?(\d\d*?)\u0020ms$/gm
        ), function (ignore, key, timeElapsed) {
            dict[key] = dict[key] || [];
            dict[key].push(Number(timeElapsed));
            return "";
        });
        console.log(dict);
        Object.keys(dict).forEach(function (key) {
            let diffSquared;
            let mean;
            let nn;
            let std;
            nn = dict[key].length;
            diffSquared = 0;
            mean = 0;
            dict[key].forEach(function (val) {
                mean += val;
            });
            mean = mean / nn;
            dict[key].forEach(function (val) {
                diffSquared += (val - mean) ** 2;
            });
            std = Math.sqrt(diffSquared / (nn - 1));
            mean = Math.round(mean);
            std = Math.round(std);
            console.log(key + "," + mean + " +/- " + std + " ms");
        });
    }
    parse(".benchmark.mkdirRecursive.txt");
    parse(".benchmark.vanilla.txt");
}());


benchmark.js
/* jslint utility2:true */
(function () {
    "use strict";
    let fs;
    let modeMkdirRecursive;
    let nn;
    let path;
    fs = require("fs");
    path = require("path");
    // init modeMkdirRecursive
    modeMkdirRecursive = process.argv[2] === "modeMkdirRecursive";
    // init nn
    nn = 1000;
    function cleanup() {
    /*
     * this function will cleanup .test dir
     */
        console.log("\ncleanup .test/");
        require("child_process").spawnSync("rm", [
            "-rf", ".test"
        ], {
            stdio: [
                "ignore", 1, 2
            ]
        });
        fs.mkdirSync(".test");
    }
    function onErrorThrow(err) {
    /*
     * this function will throw <err> if exists
     */
        if (err) {
            throw err;
        }
    }
    async function testFsOperation(operation, modeMkdir) {
    /*
     * this function will async-test <operation> with mkdir performance
     */
        let timeStart;
        async function fsOperation(filename) {
            if (modeMkdir === "fs.mkdir") {
                await new Promise(function (resolve) {
                    fs.mkdir(path.dirname(filename), {
                        recursive: true
                    }, function (err) {
                        onErrorThrow(err);
                        resolve();
                    });
                });
            }
            await new Promise(function (resolve) {
                function onError(err) {
                    onErrorThrow(err);
                    resolve();
                }
                switch (operation) {
                case "fs.appendFile":
                    fs.appendFile(filename, "hello", {
                        mkdirRecursive: modeMkdir === "mkdirRecursive"
                    }, onError);
                    break;
                case "fs.writeFile":
                    fs.writeFile(filename, "hello", {
                        mkdirRecursive: modeMkdir === "mkdirRecursive"
                    }, onError);
                    break;
                default:
                    throw new Error("invalid operation " + operation);
                }
            });
        // debug
        // console.log(operation + " " + filename);
        }
        cleanup();
        timeStart = Date.now();
        await Promise.all(Array.from(
            new Uint8Array(nn)
        ).map(function (ignore, ii) {
            let filename;
            filename = (
                modeMkdir
                ? ".test/" + ii + "/aa/bb/foo.txt"
                : ".test/" + ii + ".foo.txt"
            );
            return fsOperation(filename);
        }));
        console.log(
            nn + " * (" + operation.padEnd(17, " ") + " + "
            + String(modeMkdir || "no-mkdir").padEnd(14, " ")
            + ") - time elapsed - "
            + String(Date.now() - timeStart).padStart(4, " ") + " ms"
        );
    }
    function testFsOperationSync(operation, modeMkdir) {
    /*
     * this function will sync-test <operation> with mkdir performance
     */
        let ii;
        let timeStart;
        function fsOperation(filename) {
            if (modeMkdir === "fs.mkdirSync") {
                fs.mkdirSync(path.dirname(filename), {
                    recursive: true
                });
            }
            switch (operation) {
            case "fs.appendFileSync":
                fs.appendFileSync(filename, "hello", {
                    mkdirRecursive: modeMkdir === "mkdirRecursive"
                });
                break;
            case "fs.writeFileSync":
                fs.writeFileSync(filename, "hello", {
                    mkdirRecursive: modeMkdir === "mkdirRecursive"
                });
                break;
            default:
                throw new Error("invalid operation " + operation);
            }
        }
        cleanup();
        timeStart = Date.now();
        ii = 0;
        while (ii < nn) {
            let filename;
            filename = (
                modeMkdir
                ? ".test/" + ii + "/aa/bb/foo.txt"
                : ".test/" + ii + ".foo.txt"
            );
            fsOperation(filename);
            ii += 1;
        }
        console.log(
            nn + " * (" + operation.padEnd(17, " ") + " + "
            + String(modeMkdir || "no-mkdir").padEnd(14, " ")
            + ") - time elapsed - "
            + String(Date.now() - timeStart).padStart(4, " ") + " ms"
        );
    }
    (async function test() {
        console.log(process.versions);
        await testFsOperation("fs.writeFile", undefined);
        await testFsOperation("fs.writeFile", "fs.mkdir");
        if (modeMkdirRecursive) {
            await testFsOperation("fs.writeFile", "mkdirRecursive");
        }
        await testFsOperation("fs.appendFile", undefined);
        await testFsOperation("fs.appendFile", "fs.mkdir");
        if (modeMkdirRecursive) {
            await testFsOperation("fs.appendFile", "mkdirRecursive");
        }
        [
            "fs.writeFileSync", "fs.appendFileSync"
        ].forEach(function (operation) {
            testFsOperationSync(operation, undefined);
            testFsOperationSync(operation, "fs.mkdirSync");
            if (modeMkdirRecursive) {
                testFsOperationSync(operation, "mkdirRecursive");
            }
        });
    }());
}());
